osh/builtin_trap.py

```diff
@@ -95,12 +64,101 @@ def _GetSignalNumber(sig_spec):
 # Then hit Ctrl-C.
 
 
+class TrapState(pyos.SignalHandler):
+  """For dealing with globabl trap- and signal-related state."""
+
+  def __init__(self):
+      # type: () -> None
+      self.run_list = []  # type: List[command_t]
+      # TODO: add AddUserHook() and RemoveUserHook() for hook_nodes mutations
+      self.hook_nodes = {}  # type: Dict[str, command_t]
+      self.trap_nodes = {}  # type: Dict[int, command_t]
+      self.display = None  # type: _IDisplay
+      self.last_sig_num = 0  # type: int
+      pyos.ReserveHandlerCapacity(self.run_list)
+
+  def InitInteractiveShell(self, display, my_pid):
+    # type: (_IDisplay, int) -> None
+    """Called when initializing an interactive shell."""
+    # The shell itself should ignore Ctrl-\.
+    pyos.Sigaction(_GetSignalNumber("SIGQUIT"), SIG_IGN)
+
+    # This prevents Ctrl-Z from suspending OSH in interactive mode.
+    pyos.Sigaction(_GetSignalNumber("SIGTSTP"), SIG_IGN)
+
+    # More signals from
+    # https://www.gnu.org/software/libc/manual/html_node/Initializing-the-Shell.html
+    # (but not SIGCHLD)
+    pyos.Sigaction(_GetSignalNumber("SIGTTOU"), SIG_IGN)
+    pyos.Sigaction(_GetSignalNumber("SIGTTIN"), SIG_IGN)
+
+    # Register a callback to receive terminal width changes.
+    # NOTE: In line_input.c, we turned off rl_catch_sigwinch.
+
+    # This is ALWAYS on, which means that it can cause EINTR, and wait() and
+    # read() have to handle it
+    self.display = display
+
+    if mylib.PYTHON:
+      pyos.Sigaction(SIGWINCH, self)
+
+  def AddUserTrap(self, sig_num, node):
+    # type: (int, command_t) -> None
+    """For user-defined handlers registered with the 'trap' builtin."""
+
+    self.trap_nodes[sig_num] = node
+    if sig_num == SIGWINCH:
+      if mylib.PYTHON:
+        assert self.display is not None
+        pyos.Sigaction(sig_num, self)
+    else:
+      pyos.Sigaction(sig_num, self)
+    # TODO: SIGINT is similar: set a flag, then optionally call user handler
+
+  def RemoveUserTrap(self, sig_num):
+    # type: (int) -> None
+    """For user-defined handlers registered with the 'trap' builtin."""
+    # Restore default
+    if sig_num == SIGWINCH:
+      if mylib.PYTHON:
+        self.trap_nodes[sig_num] = None
+    else:
+      pyos.Sigaction(sig_num, SIG_DFL)
+    if sig_num in self.trap_nodes:
+      del self.trap_nodes[sig_num]
+    # TODO: SIGINT is similar: set a flag, then optionally call user handler
+
+  def TakeRunList(self):
+      # type: () -> List[command_t]
+      new_run_list = []  # type: List[command_t]
+      pyos.ReserveHandlerCapacity(new_run_list)
+      ret = self.run_list
+      self.run_list = new_run_list
+      return ret
```

Comment:
This gets transpiled to the following, which _should_ help with one of the races you pointed out @andychu. Assuming pointer assignment on the underlying arch is atomic  
```
List<syntax_asdl::command_t*>* TrapState::TakeRunList() {
  List<syntax_asdl::command_t*>* new_run_list = nullptr;
  List<syntax_asdl::command_t*>* ret = nullptr;
  StackRoots _roots({&new_run_list, &ret});

  new_run_list = Alloc<List<syntax_asdl::command_t*>>();
  pyos::ReserveHandlerCapacity(new_run_list);
  ret = this->run_list;
  this->run_list = new_run_list;
  return ret;
}
```